docs/design/014-did-doc/README.md

```diff
@@ -0,0 +1,590 @@
+1. publish new DID Doc to microledger
+1. add second key to the DID DOC from a different edge device
+1. rotate a key from an edge device
+1. change authentication/authorization policies from one of the devices
+1. change service endpoints from an edge device
+1. support of multiple endpoints, each working with multiple edge devices
+1. know the recent key for this edge device to be used for signature and A2A communication using this edge device
+1. know if the key on the edge device can be used to perform modification of DID Doc on the ledger/microledger
+1. have multiple keys associated with same DID on edge device
+1. know the most recent DID DOC for the other side of communication
+1. communicate if other side of communication uses a different did method:
+    - For example, communication between entities from STN and SLN
+    - For example, Communication between Indy/Sovrin and UPort
+1. know what key(s) to use for signing/packing/A2A
+1. know what key(s) to use for verification/unpacking/A2A
+1. be able to publish existing DID Doc (maybe not as is) into a different ledger/microledger
+1. be able to re-use existing crypto material in the new approach
+1. communicate with clients using old (existing) approach
+1. use different keys for A2A and client-to-ledger communication
+
+### Create new DID and DID DOC
+
+![Alt](./create-did-ddoc.svg "Create new DID and DID DOC")
+
+### Create key on 2nd Edge Device
+
+![Alt](./create-key-2nd-device.svg "Create key on 2nd Edge Device")
+
+### Rotate a key by an Edge Device who is authenticated to do the update
+
+![Alt](./rotate-key-with-auth.svg "Rotate a key by an Edge Device who is authenticated to do the updatee")
+
+### Rotate a key by an Edge Device who is not authenticated to do the update
+
+![Alt](./rotate-key-without-auth.svg "Rotate a key by an Edge Device who is not authenticated to do the updatee")
+
+### A2A Communication
+
+![Alt](./a2a.svg "A2A Communication")
+
+## Identity structure
+
+On Indy SDK level DID-related identity has the following structure:
+
+- DID Method
+- Id string
+- Keys (private and public) on different edge devices
+  - Private keys are not shared between devices (the only possible exception is backup)
+- Authorization, authentication and discovery meta information:
+  - Public copy on ledger as DID DOC
+  - Optional local copy or cache
+
+## Design principles
+
+1. Each edge device stores the following information for every owned DID:
+    - Fully-qualified Did - Identifies data in storages
+    - Private and public keys associated with the DID from this edge device - in Wallet
+2. Each edge device can optionally store cached local copy of DID Doc for owned DID:
+    - The current DID Doc copy is used for signing/pack.
+    - The current DID Doc copy is synchronized between devices only through the Ledger (Public or microledger).
+    - DID Doc is never sent from one device to another.
```

Comment:
In case of microledger?

Reply:
@lovesh @dhh1128 @ashcherbakov

> In case of microledger?

It isn't important how DID Doc is shared. There is some storage (public ledger, micro ledger, centralized storage, some ledger between cloud agents) that all connected devices have access and this storage is used to perform DID Doc synchronization. Otherwise you can easy get DID Doc out of think that can cause security and communication problems.

Reply:
Yes, we propose to separate at least the following layers/protocols:
1) Sub-protocol of how DID DOC created and updated by agents
2) Sub-protocol of how DID DOC updates are ordered and synchronized between Alice's devices .
Options: public ledger, BFT-like microledger, CFT-like microledger, Centralized approach, etc.
3) Sub-protocol of how and when Bob get the latest Alice's DID DOC.
We will update the docs with this thinking.

Reply:
I think the impacts of the storage solution that is being discussed will play into this discussion. With regards to microledger sharing, I think within a domain (all of alice's agents) we can rely on  a storage sync layer to handle it. With updates across domains (e.g. bob publishes update to alice) we would likely use DIDComms to communicate the update, and alice would store it locally in her storage sync layer.

Worth noting storage syncing is still very early thinking, so this discussion will be an important use case when we're looking to design it.